crates/apollo-router-core/src/query_planner/caching_query_planner.rs

```diff
@@ -37,25 +37,75 @@ impl<T: QueryPlanner> QueryPlanner for CachingQueryPlanner<T> {
         operation: Option<String>,
         options: QueryPlanOptions,
     ) -> Result<Arc<QueryPlan>, QueryPlannerError> {
-        if let Some(value) = self
-            .cached
-            .lock()
-            .await
-            .get(&(query.clone(), operation.clone(), options.clone()))
-            .cloned()
-        {
+        let mut locked_cache = self.cached.lock().await;
+        let key = (query.clone(), operation.clone(), options.clone());
+        if let Some(value) = locked_cache.get(&key).cloned() {
             return value;
         }
 
-        let value = self
-            .delegate
-            .get(query.clone(), operation.clone(), options.clone())
-            .await;
-        self.cached
-            .lock()
-            .await
-            .insert((query, operation, options), value.clone());
-        value
+        // Holding a lock across the delegated get is a bad idea because
+        // the delegate get() calls into v8 for processing of the plan.
+        // This would block all other get() requests for a potentially
+        // long time.
+        // Alternatively, if we don't hold the lock, there is a risk
+        // that we will do the work multiple times. This is also
+        // sub-optimal.
+
+        // To work around this, we keep a list of keys we are currently
+        // processing in the delegate. If we try to get a key on this
+        // list, we block and wait for it to complete and then retry.
+        //
+        // This is more complex than either of the two simple
+        // alternatives but succeeds in providing a mechanism where each
+        // client only waits for uncached QueryPlans they are going to
+        // use AND avoids generating the plan multiple times.
+
+        drop(locked_cache);
+
+        let mut locked_wait_map = self.wait_map.lock().await;
+
+        match locked_wait_map.get_mut(&key) {
+            Some(waiter) => {
+                // Register interest in key
+                let mut receiver = waiter.subscribe();
+                drop(locked_wait_map);
+                let msg = receiver.recv().await?;
+                assert_eq!(msg, key);
+                let mut locked_cache = self.cached.lock().await;
```

Comment:
repeating what I said in the outdated comment:
It might be useful to note here that in extreme cases, between the moment we receive the notification and the moment we check the cache, the value could have been evicted from the cache
so we could instead get the value from the receiver directly and avoid that race ?

Reply:
Great idea. I've done this.